// @youcan/theme@2.0.0/dist/cli/commands/theme/delete.js
import { Session, Http, Env, Tasks } from '@youcan/cli-kit';
import { ThemeCommand } from '../../../util/theme-command.js';

class Delete extends ThemeCommand {
    static description = 'Select remote development themes to delete';
    async run() {
        await Session.authenticate(this);
        const { dev: themes } = await Http.get(`${Env.apiHostname()}/themes`);
        if (!themes.length) {
            return this.output.info('You have no remote dev themes');
        }
        const choices = themes.map(t => ({ title: `${t.name} (${t.id})`, value: t.id }));
        const { identifiers } = await this.prompt({
            choices,
            type: 'multiselect',
            name: 'identifiers',
            message: 'Select themes to delete',
        });
        const tasks = identifiers.map((id) => {
            const theme = themes.find(t => t.id === id);
            return {
                title: `Deleting '${theme.name} (${theme.id})'...`,
                async task() {
                    await Http.post(`${Env.apiHostname()}/themes/${theme.id}/delete`);
                },
            };
        });
        await Tasks.run({}, tasks);
        this.output.info(`Done! ${identifiers.length} themes deleted.`);
    }
}

export { Delete as default };
