// @youcan/theme@2.0.0/dist/cli/commands/theme/dev.js
import { Session, Tasks, Http, Env } from '@youcan/cli-kit';
import { ThemeCommand } from '../../../util/theme-command.js';
import { load } from '../../../util/theme-loader.js';
import ThemeWorker from '../../services/dev/worker.js';

class Dev extends ThemeCommand {
    static description = 'Start a theme development server and preview your changes';
    async run() {
        const theme = await load();
        await Session.authenticate(this);
        const context = await Tasks.run({ cmd: this, workers: [] }, [
            {
                title: 'Fetching theme info...',
                async task(ctx) {
                    const res = await Http.get(`${Env.apiHostname()}/me`);
                    ctx.store = {
                        slug: res.slug,
                        domain: res.domain,
                    };
                },
            },
            {
                title: 'Preparing dev processes...',
                async task(ctx) {
                    ctx.workers = [
                        new ThemeWorker(ctx.cmd, ctx.store, theme),
                    ];
                    await Promise.all(ctx.workers.map(async (w) => await w.boot()));
                },
            },
        ]);
        await Promise.all(context.workers.map(async (w) => await w.run()));
    }
}

export { Dev as default };
